shared/src/publication/Link.ts

```diff
@@ -0,0 +1,330 @@
+    link.properties = link.properties
+      ? link.properties?.add(properties)
+      : new Properties(properties);
+    return link;
+  }
+
+  /**
+   * Creates a [Locator] from a reading order [Link].
+   */
+  public get locator(): Locator {
+    let parts = this.href.split('#');
+    return new Locator({
+      href: parts.length > 0 && parts[0] !== undefined ? parts[0] : this.href,
+      type: this.type ?? '',
+      title: this.title,
+      locations: new LocatorLocations({
+        fragments: parts.length > 1 && parts[1] !== undefined ? [parts[1]] : [],
+      }),
+    });
+  }
+}
+
+/**
+ * Parses multiple JSON links into an array of Link.
+ */
+export class Links {
+  public items: Array<Link>;
+
+  /**
+   * Creates a [Links].
+   */
+  constructor(items: Array<Link>) {
+    this.items = items;
+  }
+
+  /**
+   * Creates a list of [Link] from its RWPM JSON representation.
+   */
+  public static deserialize(json: any): Links | undefined {
+    if (!(json && json instanceof Array)) return;
+    return new Links(
+      json
+        .map<Link>(item => Link.deserialize(item) as Link)
+        .filter(x => x !== undefined)
+    );
+  }
+
+  /**
+   * Serializes an array of [Link] to its RWPM JSON representation.
+   */
+  public serialize(): any {
+    return this.items.map(x => x.serialize());
+  }
+
+  /** Finds the first link with the given relation. */
+  public findWithRel(rel: string): Link | undefined {
+    const predicate = (el: Link) => el.rels && el.rels.has(rel);
+    return this.items.find(predicate);
+  }
+
+  /** Finds all the links with the given relation. */
+  public filterByRel(rel: string): Array<Link> {
+    const predicate = (el: Link) => el.rels && el.rels.has(rel);
+    return this.items.filter(predicate);
+  }
+
+  /** Finds the first link matching the given HREF. */
+  public findWithHref(href: string): Link | undefined {
+    const predicate = (el: Link) => el.href === href;
+    return this.items.find(predicate);
+  }
+
+  /** Finds the index of the first link matching the given HREF. */
+  public findIndexWithHref(href: string): number {
+    const predicate = (el: Link) => el.href === href;
+    return this.items.findIndex(predicate);
+  }
+
+  /** Finds the first link matching the given media type. */
+  public findWithMediaType(mediaType: string): Link | undefined {
+    const predicate = (el: Link) => el.mediaType.matches(mediaType);
+    return this.items.find(predicate);
+  }
+
+  /** Finds all the links matching the given media type. */
+  public filterByMediaType(mediaType: string): Array<Link> {
+    const predicate = (el: Link) => el.mediaType.matches(mediaType);
+    return this.items.filter(predicate);
+  }
+
+  /** Finds all the links matching any of the given media types. */
+  public filterByMediaTypes(mediaTypes: Array<string>): Array<Link> {
+    const predicate = (el: Link) => {
+      for (const mediaType of mediaTypes) {
+        return el.mediaType.matches(mediaType);
```

Comment:
I'm pretty sure this will only check the first element of the array. You need to only `return` if it's found.

```suggestion
        if (el.mediaType.matches(mediaType)) {
          return true;
        }
```

Reply:
good point. fixed. thanks